### ISSUE TITLE
Calling mono's Navigation.GetSimplePath() crashes the game

### ISSUE THREAD

**Issue by tagcup** [godotengine/godot#11957]
Attached a sample project.

[mononav.zip](https://github.com/godotengine/godot/files/1367285/mononav.zip)

**Comment by heikytr** [comment]
The issue here isn't the `GetSimplePath()` method, but actually the `GD.print(path);` line, I don't know exactly why it's crashing but it crashes when it gets to this line. You can try using:
```C#
for (int i = 0; i < path.Length; i++)
{
    GD.print(path[i]);
}
```

and it will work just fine, except that it shows us that `GetSimplePath()` returns only the **X** axis, a problem for another issue.

**Comment by tagcup** [comment]
No, the crash *does* happen without the print. I added it later on, but you can simply remove it to confirm that.

That being said, if you add that for loop, it actually no longer crashes, which is quite odd.

**Comment by heikytr** [comment]
Just confirmed and without the print line it does crash, odd indeed because I was testing earlier with another project and it works fine without the print or the for loop. Let me clean up the project and I will upload it here.

**Comment by tagcup** [comment]
Just speculating here but it might be something related to C#'s garbage collector acting aggressively

**Comment by heikytr** [comment]
Sorry if it took too long, I basically just rewrote it using as example the nav mesh demo, it works.

[MonoNavTests.zip](https://github.com/godotengine/godot/files/1367556/MonoNavTests.zip)

**Comment by tagcup** [comment]
You're passing the result of GetSimplePath to

> pathList = new List<Vector3>(pathArray.Reverse().ToList());

I wonder, does it crash if you don't make use of pathList at all? (sorry, can't test it right now)

**Comment by neikeq** [comment]
This is a marshalling error. The following macro expands to `sizeof(real_t *)` but it should be `sizeof(real_t) * 3`:

https://github.com/godotengine/godot/blob/bf12b459864f89d056588c67cd90d0578681146b/modules/mono/mono_gd/gd_mono_marshal.cpp#L783

I'll fix when I get home.

**Comment by tagcup** [comment]
Just tried with the current master, the example project I attached in the original poster is still crashing actually

**Comment by tagcup** [comment]
[This](https://gist.github.com/tagcup/b15a68b61529655dc88aad59c52fe8ed) is what I see in stderr, and the contents of the log file are as in #11929

**Comment by tagcup** [comment]
BTW, while the crash message is different, the way to trigger the toggle the example project remains the same: the crash is still gone after commenting out the line with GetSimplePath

**Comment by tagcup** [comment]
Can we reopen this issue?
I can open a new one, but the description and attached example project would be exactly the same.

**Comment by neikeq** [comment]
@tagcup is this crash with mononav.zip or MonoNavTests.zip?

**Comment by neikeq** [comment]
I can't reproduce it, but does #12000, by any chance, fix it?

**Comment by tagcup** [comment]
> I can't reproduce it

Wow that's odd. Sure, I'll give it another try later today this evening, and I want to do a clean clone before compiling and also remove ~/.godot entirely

**Comment by neikeq** [comment]
Actually, now I managed to reproduce a failed assertion on exit. Commenting the exported texture field fixes it. Is that your issue?

**Comment by tagcup** [comment]
No, although that sounds like something related to [this](https://github.com/godotengine/godot/issues/11955#issuecomment-335048025).

In my case, commenting out the line with GetSimplePath fixed the crash (until yesterday).

**Comment by tagcup** [comment]
With crash, I mean a segfault when trying to run the game at start, not a failed assertion on exit

**Comment by heikytr** [comment]
@tagcup Using the last commit, I also got the same error with the mononav.zip, but only after I close the running scene, which `Commenting the exported texture field` fixed here, the crash after **GetSimplePath** is totally gone for me.

@neikeq  The #12000 Actually fixed the issue that **GetSimplePath** was returning only the **X** axis or the `raw[0] = elem.x;` from Vector3.

**Comment by tagcup** [comment]
I can confirm that this is fixed with the current master and the path seems to be correct.

Should I open a new issue related to the failed assertion on exit, or is #11955 enough?

**Comment by neikeq** [comment]
All right, closing then. Regarding the exported texture field crash, open a new one, it's not related to #11955.
